libbpf-tools/sysinjector.c

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+/* Copyright 2021 Sony Group Corporation */
+#include <argp.h>
+#include <errno.h>
+#include <signal.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+
+#include "trace_helpers.h"
+#include "sysinjector.h"
+#include "sysinjector.skel.h"
+
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+static struct prog_env {
+	pid_t pid;
+	char *comm;
+	int comm_len;
+	int retval;
+	char *syscall;
+} env;
+
+const char *argp_program_version = "sysinjector 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+static const char args_doc[] = "SYSCALL";
+static const char program_doc[] =
+"Modify a return value of a syscall\n"
+"\n"
+"Usage: sysinjector [-h] [-p PID] [-c COMM] [-r RETVAL] SYSCALL\n"
+"\v"
+"Examples:\n"
+"  ./sysinjector -r -12 -p 1111 __x64_sys_mknodat            # return ENOMEM\n"
+"  ./sysinjector -r -22 -c sleep __x64_sys_clock_nanosleep   # return EINVAL\n"
+;
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Process ID to modify"},
+	{ "comm",  'c', "COMM",  0, "Trace this comm only" },
+	{ "retval", 'r', "RETVAL", 0, "Modified return value" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help"},
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	struct prog_env *env = state->input;
+	long pid;
+	int len, retval;
+
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'p':
+		errno = 0;
```

Comment:
You don't need to set errno explicitly, do you ?

Reply:
_strtol()_  can legitimately return 0, LONG_MAX, or LONG_MIN on both success and failure. So, this code should set errno to 0 before calling _strtol()_ .
https://man7.org/linux/man-pages/man3/strtol.3.html